Non-pruning timemanagement for self-play

Ttl opened the pull request:
Adds support for timemanagement code in leelaz and autogtp that doesn't prune moves and only cuts off the search if the second best move can't overtake the best one with the remaining visits.

The visit distribution with timemanagement is very close to without it and it generates very similar training data. On average less visits are needed per move speeding up the game generation.

More discussion in #1416

Eddh commented:
For what it's worth I don't like these kind of modifications because it opens up new possibilities for the net to make its accuracy higher without actually learning anything useful. In this case the net could "cheat" by learning to recognize situations in which the search will be cut short in order to get a better policy loss. It would not help it in real games, and yet filters, layers might be wasted on that.

Ttl commented:
The visit distribution is very similar to one played at fixed visits. I don't see how knowing that the position got less visits than maximum would give advantage in prediction the policy distribution. Can you give an example?

gcp commented:
>In this case the net could "cheat" by learning to recognize situations in which the search will be cut short in order to get a better policy loss.

How would this happen?

gcp commented on src/Leela.cpp:
```diff
@@ -277,6 +277,8 @@ static void parse_commandline(int argc, char *argv[]) {
         } else if (tm == "fast") {
             cfg_timemanage = TimeManagement::FAST;
+        } else if (tm == "simple") {
+            cfg_timemanage = TimeManagement::SIMPLE;
```
We already make quite a few things conditional on noise/random selection so this could as well.

I understand that above you've added a possibility for the server to add arbitrary options to enable this? That has the disadvantage that it requires all clients to be switched. (Unless AutoGTP behaves correctly with unknown options).

Ttl replied:
I thought that server can give different commands to different versions. Would you rather have it default to simple?

I think that autogtp ignores unknown options. It parses json to ```map<QString, QString>``` and only checks keys for commands it expects.

gcp replied:
>I thought that server can give different commands to different versions. 

Ha, good point.

>Would you rather have it default to simple?

We could do that if we detect noise is enabled. BTW. What about DONT_PRUNE or NO_PRUNING instead of SIMPLE?

Maybe FAST should have been named VERY_EARLY_EXIT or something. The naming here kinda sucks right now.

Eddh commented:
OK maybe I misunderstood something I will read in more detail.

marcocalignano commented on autogtp/Management.cpp:
```diff
@@ -269,6 +269,8 @@ QString Management::getOptionsString(const QJsonObject &opt, const QString &rnd)
         options.append(" -s " + rnd + " ");
     }
+    // Additional options specified by server
+    options.append(getOption(opt, "other_options", "", ""));
```
What is this for?

Ttl replied:
Can be used to give arbitrary command line options to leelaz. For example timemanage, randomvisits, randomtemp. However I just removed it from this pull request since now non_pruning timemanagement is enabled by default.

Ttl commented:
Now enabled by default when noise is enabled. There is legitimate uses for using "-m" option in match games to get more random opening so I changed auto to look for noise option instead. Renamed simple to non_pruning. I didn't change FAST because the command line parameter for it is fast.

marcocalignano commented:
@Ttl  If you change things with the option you have to be aware of the save restore procedure, you have to store all the options  that you gave Leelaz that is way I would like not to have a general option flag.

Ishinoshita commented:
Hi @Ttl. Since I did some analysis of LZ177 training data that turned out to revolve around time_mgt_pruning without knowing it, let me share it:
[LZ177 training data - Effect of time management pruning on policy.pptx](https://github.com/gcp/leela-zero/files/2485012/LZ177.training.data.-.Effect.of.time.management.pruning.on.policy.pptx)

I can definitively see statistically significant differences in the distribution of policy targets, with  timemanagment pruning On or Off, but I cannot tell whether this is good or bad for training and if there is an issue with that option for self-play. Thus, I share the comparison, without any comment.

Just to illustrate directly in this thread:

![image](https://user-images.githubusercontent.com/37498331/47049068-828a3880-d19c-11e8-831f-d41ae17e21fa.png)

![image](https://user-images.githubusercontent.com/37498331/47049091-96ce3580-d19c-11e8-90e9-6ce637ffaef9.png)

Let me clarify that I did this comparison after removing *all games* affected by any of the new weirdnesses I found and I gonna mention in a new post on LZ177 traing data.

For the methodology, see [my git](https://github.com/Ishinoshita/Leela-Zero-Training-Data-Viewer).
